profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/TimerCreateCpuProfiler.cpp

```diff
@@ -0,0 +1,216 @@
+    CallstackProvider callstackProvider) noexcept
+    :
+    _pSignalManager{pSignalManager}, // put it as parameter for better testing
+    _pManagedThreadsList{pManagedThreadsList},
+    _pProvider{pProvider},
+    _callstackProvider{std::move(callstackProvider)},
+    _serviceState{ServiceState::Initialized},
+    _samplingInterval{pConfiguration->GetCpuProfilingInterval()}
+{
+    Log::Info("Cpu profiling interval: ", _samplingInterval.count(), "ms");
+
+    Instance = this;
+}
+
+TimerCreateCpuProfiler::~TimerCreateCpuProfiler()
+{
+    Stop();
+}
+
+void TimerCreateCpuProfiler::RegisterThread(std::shared_ptr<ManagedThreadInfo> threadInfo)
+{
+    if (_serviceState != ServiceState::Started)
+    {
+        return;
+    }
+
+    auto tid = threadInfo->GetOsThreadId();
+    Log::Debug("Creating timer for thread ", tid);
+
+    struct sigevent sev;
+    sev.sigev_value.sival_ptr = nullptr;
+    sev.sigev_signo = _pSignalManager->GetSignal();
+    sev.sigev_notify = SIGEV_THREAD_ID;
+    ((int*)&sev.sigev_notify)[1] = tid;
+
+    // Use raw syscalls, since libc wrapper allows only predefined clocks
+    clockid_t clock = ((~tid) << 3) | 6; // CPUCLOCK_SCHED | CPUCLOCK_PERTHREAD_MASK thread_cpu_clock(tid);
+    int timerId;
+    if (syscall(__NR_timer_create, clock, &sev, &timerId) < 0)
+    {
+        Log::Error("Call to timer_create failed for thread ", tid);
+        return;
+    }
+
+    auto oldTimerId = threadInfo->SetTimerId(timerId);
+
+    // In case of SSI:
+    // There is a race when Start() and RegisterThread :
+    // The thread can be added while lookping over the in the managed threads list
+    // and a call being made to RegisterThread
+    // In that case, just keep the first timer id and delete the other
+    if (oldTimerId != -1)
+    {
+        // delete the newly created timer
+        syscall(__NR_timer_delete, timerId);
+        threadInfo->SetTimerId(oldTimerId);
+        return;
+    }
+
+    std::int32_t _interval = std::chrono::duration_cast<std::chrono::nanoseconds>(_samplingInterval).count();
+    struct itimerspec ts;
+    ts.it_interval.tv_sec = (time_t)(_interval / 1000000000);
+    ts.it_interval.tv_nsec = _interval % 1000000000;
+    ts.it_value = ts.it_interval;
+    syscall(__NR_timer_settime, timerId, 0, &ts, nullptr);
+}
+
+void TimerCreateCpuProfiler::UnregisterThread(std::shared_ptr<ManagedThreadInfo> threadInfo)
+{
+    if (_serviceState != ServiceState::Started)
+    {
+        Log::Debug("timer_create-based cpu profiler is not started. Cannot unregister thread.");
+        return;
+    }
+
+    auto timerId = threadInfo->SetTimerId(-1);
+
+    Log::Debug("Unregister timer for thread ", threadInfo->GetOsThreadId());
+    syscall(__NR_timer_delete, timerId);
+}
+
+const char* TimerCreateCpuProfiler::GetName()
+{
+    return "timer_create-based Cpu Profiler";
+}
+
+bool TimerCreateCpuProfiler::Start()
+{
+    if (_serviceState.exchange(ServiceState::Started) == ServiceState::Started)
```

Comment:
As discussed, being Started before iterating the ManageThreadList ensure that we won't miss any new thread created during the list iterations.
However, it is possible to add a time to a thread that was/is about to be destroyed.
Check if adding a lock in Register/Unregister/Start could solve these race conditions